src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1753 @@
+    this(config, IdentifierRules.DEFAULT);
+  }
+
+  /**
+   * Create a new dialect instance with the given connector configuration.
+   *
+   * @param config                 the connector configuration; may not be null
+   * @param defaultIdentifierRules the default rules for identifiers; may be null if the rules are
+   *                               to be determined from the database metadata
+   */
+  protected GenericDatabaseDialect(
+      AbstractConfig config,
+      IdentifierRules defaultIdentifierRules
+  ) {
+    this.config = config;
+    this.defaultIdentifierRules = defaultIdentifierRules;
+    this.jdbcUrl = config.getString(JdbcSourceConnectorConfig.CONNECTION_URL_CONFIG);
+    if (!(config instanceof JdbcSinkConfig)) {
+      mapNumerics = config.getBoolean(JdbcSourceTaskConfig.NUMERIC_PRECISION_MAPPING_CONFIG);
+      catalogPattern = config.getString(JdbcSourceTaskConfig.CATALOG_PATTERN_CONFIG);
+      schemaPattern = config.getString(JdbcSourceTaskConfig.SCHEMA_PATTERN_CONFIG);
+      tableTypes = new HashSet<>(config.getList(JdbcSourceTaskConfig.TABLE_TYPE_CONFIG));
+    } else {
+      mapNumerics = JdbcSourceTaskConfig.NUMERIC_PRECISION_MAPPING_DEFAULT;
+      catalogPattern = JdbcSourceTaskConfig.CATALOG_PATTERN_DEFAULT;
+      schemaPattern = JdbcSourceTaskConfig.SCHEMA_PATTERN_DEFAULT;
+      tableTypes = new HashSet<>(Arrays.asList(JdbcSourceTaskConfig.TABLE_TYPE_DEFAULT));
+    }
+  }
+
+  @Override
+  public Connection getConnection() throws SQLException {
+    // These config names are the same for both source and sink configs ...
+    String username = config.getString(JdbcSourceConnectorConfig.CONNECTION_USER_CONFIG);
+    Password dbPassword = config.getPassword(JdbcSourceConnectorConfig.CONNECTION_PASSWORD_CONFIG);
+    Properties properties = new Properties();
+    if (username != null) {
+      properties.setProperty("user", username);
+    }
+    if (dbPassword != null) {
+      properties.setProperty("password", dbPassword.value());
+    }
+    properties = addConnectionProperties(properties);
+    Connection connection = DriverManager.getConnection(jdbcUrl, properties);
+    if (jdbcDriverInfo == null) {
+      jdbcDriverInfo = createJdbcDriverInfo(connection);
+    }
+    connections.add(connection);
+    return connection;
+  }
+
+  @Override
+  public void close() {
+    while (!connections.isEmpty()) {
+      Connection conn = connections.poll();
+      if (conn != null) {
+        try {
+          conn.close();
+        } catch (SQLException e) {
+          log.warn("Error while closing connection to {}", jdbcDriverInfo, e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public boolean isConnectionValid(
+      Connection connection,
+      int timeout
+  ) throws SQLException {
+    if (jdbcDriverInfo().jdbcMajorVersion() >= 4) {
+      return connection.isValid(timeout);
+    }
+    // issue a test query ...
+    String query = checkConnectionQuery();
```

Comment:
Here's an example of how we can't simply call a JDBC 4 method, since they won't work on drivers that are only at JDBC 2 or 3.